### PR TITLE
Fix infinite loop on CRC error

### LINKED ISSUE #36074: LoRaWAN: sx126x: infinite loop on CRC error
**Describe the bug**

When receiving a packet on LoRaWAN that contains a CRC error, the next attempt to use the modem will result in an infinite loop in `SX126xWaitOnBusy`.

This is due to the RX interrupt handler putting the modem into sleep mode, followed by the CRC handler setting the module state to `MODE_STDBY_RC` without actually changing the hardware state:
https://github.com/zephyrproject-rtos/loramac-node/blob/2cee5f7295ff0ff804bf06fea5de006bc7cd121e/src/radio/sx126x/radio.c#L1154

This has been resolved upstream in the following commit: https://github.com/Lora-net/LoRaMac-node/commit/fbb6a97774e63643a11e4d3334458efd7879867b

**To Reproduce**

The issue can be reproduced by simulating a CRC error in `RadioIrqProcess`:
```
        uint16_t irqRegs = SX126xGetIrqStatus( );
        SX126xClearIrqStatus( IRQ_RADIO_ALL );

        static uint8_t cnt = 0;
        if (irqRegs & IRQ_RX_DONE) {
            if (++cnt >= 5) {
                irqRegs |= IRQ_CRC_ERROR;
                cnt = 0;
            }
        }
        LOG_INF("IRQ: 0x%04X %d", irqRegs, cnt);
```
This will generate a CRC error on the 5th time the modem receives, triggering the infinite loop.

**Expected behavior**
CRC errors should just be dropped, not cause unrecoverable loops.

**Impact**
Depending on which context the application is calling `lorawan_send`, this will either stop LoRaWAN working, lock up the main loop, or result in a watchdog reboot.

**Logs and console output**
The following show the modem being set into sleep mode, then the mode incorrectly being set to `STDBY_RC`.
```
[00:00:14.901,641] <inf> sx126x: SetOperatingMode: TX (4)
[00:00:15.231,903] <inf> app: IRQ: 0x0001
[00:00:15.231,903] <inf> sx126x: SetOperatingMode: STDBY_RC (1)
[00:00:15.231,994] <inf> sx126x: SetOperatingMode: SLEEP (0)
[00:00:16.222,229] <inf> sx126x: Waiting for device...
[00:00:16.223,297] <inf> sx126x: Device ready
[00:00:16.223,419] <inf> sx126x: SetOperatingMode: STDBY_RC (1)
[00:00:16.226,654] <inf> sx126x: SetOperatingMode: RX (5)
[00:00:16.265,930] <inf> app: IRQ: 0x0004
[00:00:16.404,602] <inf> app: IRQ: 0x0010
[00:00:16.568,511] <inf> app: IRQ: 0x0042
[00:00:16.568,511] <inf> sx126x: SetOperatingMode: STDBY_RC (1)
[00:00:16.569,305] <inf> sx126x: SetOperatingMode: SLEEP (0)
[00:00:16.571,441] <err> lorawan: McpsIndication failed : Address fail
[00:00:16.571,472] <inf> sx126x: SetOperatingMode: STDBY_RC (1)
[00:00:16.571,502] <inf> lorawan_controller: Unconfirmed send succeeded
....
```
**Environment (please complete the following information):**
 - Zephyr v2.5
 - Zephyr v2.6

### PR DESCRIPTION
**Pull request by JordanYates**
This PR cherry-picks two commits from the upstream repository that resolve https://github.com/zephyrproject-rtos/zephyr/issues/36074.

The core of these commits is that only one of the CRC Error and RX done handlers are run, instead of the RX done running and then the CRC Error running.

This is done as a dedicated PR (instead of waiting for #9) to allow these fixes to be used directly by Zephyr v2.5 and Zephyr v2.6 without requiring additional changes in the Zephyr drivers.

### CODE REVIEW

**Comment by tagunil** [comment]
I have a feeling that we should reorganize this repository somehow if we want to support older Zephyr versions. Maybe something like backport branches per version. @Mani-Sadhasivam, @carlescufi, do we have any such cases for other modules?

**Comment by JordanYates** [comment]
This is a general problem with the modules based on third-party repositories.
If you track the third-party main: and have `zephyr_vX.Y` branches:
* where does the latest zephyr development branch point?
* Who is responsible for and how do the commits per version get reviewed?
* The Zephyr glue code commits will exist on each branch (not a huge problem IMO)

If you do the current setup:
* How do you apply bugfixes to previous releases (especially once `main` has moved on)?
* Is someone supposed to review the unholy abominations of the update PR's? (#9)

https://docs.zephyrproject.org/latest/guides/modules.html#allowed-practices doesn't really allow for the first option, even though it most resembles the core Zephyr repository.

**Comment by carlescufi** [comment]
I suggest we stay with the current setup and then:

>How do you apply bugfixes to previous releases (especially once main has moved on)?

You create a `zephy-vX.Y-branch` and post the fixes there.

>Is someone supposed to review the unholy abominations of the update PR's? (#9)

What is the alternative?

**Comment by carlescufi** [comment]
@JordanYates @tagunil @Mani-Sadhasivam we give a lot of flexibility to module repo maintainers in order to resolve problems like the ones described here. I personally do not have a particular preference other than "let us do whatever is the simplest". In general we haven't had severe issues with other similar modules in the past (see openthread or tf-m, cc @rlubos and @ioannisg), but I will defer to you all to decide what to do in this case. In general however, let us figure out how to move forward for the current tip of `main` in the zephyr tree and, if fixes are required for an already released Zephyr version, branch out at that point and not earlier.

**Review by carlescufi — APPROVED** [review]

**Comment by tagunil** [comment]
I think the cleanest way to move forward at the moment would be to:
1. Rename the `master` branch to `zephyr_v2.6` and add these backported fixes to it;
2. Create the new `zephyr` branch from the latest stable upstream release (v4.5.2) with minimal Zephyr-related changes on top of that;
3. Move other Zephyr-specific stuff to `modules/` subdirectory of the main Zephyr tree.

@carlescufi, @JordanYates, @Mani-Sadhasivam, what do you think about that?

**Comment by JordanYates** [comment]
>> Is someone supposed to review the unholy abominations of the update PR's? (#9)

> What is the alternative?

The upstream maintainers have already reviewed and approved those commits.
Any problems should IMO be addressed upstream, not as modifications to months old commits on our own tree.

**Comment by JordanYates** [comment]
> 1. Rename the master branch to zephyr_v2.6 and add these backported fixes to it;
> 2. Create the new zephyr branch from the latest stable upstream release (v4.5.2) with minimal Zephyr-related changes on top of that;
> 3. Move other Zephyr-specific stuff to modules/ subdirectory of the main Zephyr tree.

I think this is a reasonable solution that most importantly can also be repeated for later releases we want to upgrade to.

**Comment by Mani-Sadhasivam** [comment]
> I think the cleanest way to move forward at the moment would be to:
> 
> 1. Rename the `master` branch to `zephyr_v2.6` and add these backported fixes to it;
> 2. Create the new `zephyr` branch from the latest stable upstream release (v4.5.2) with minimal Zephyr-related changes on top of that;
> 3. Move other Zephyr-specific stuff to `modules/` subdirectory of the main Zephyr tree.

Sounds good to me! Let's merge these two commits first into `master` then checkout `zephyr-v2.6`. Later work on the update PR and merge it into `master`.

**Review by Mani-Sadhasivam — COMMENTED** [review]
LGTM

**Pull request merged**
